neps/nep-0452.md

```diff
@@ -0,0 +1,220 @@
+   /// indicating that all those assets will be sent to the claiming account (either new or   
+   /// existing) when the key is successfully used.
+   ft_data: FTData[] | null,
+
+   /// ... other types can be introduced and the standard is easily extendable.
+}
+
+
+/// Data outlining a specific Non-Fungible Token that should be sent to the claiming account /// (either new or existing) when a key is successfully used.
+type NFTData = {
+   /// the id of the token to transfer
+   token_id: string,
+   /// The valid NEAR account indicating the Non-Fungible Token contract.
+   contract_id: string
+}
+
+
+/// Data outlining Fungible Tokens that should be sent to the claiming account 
+/// (either new or existing) when a key is successfully used.
+type FTData = {
+   /// The number of tokens to transfer, wrapped in quotes and treated
+   /// like a string, although the number will be stored as an unsigned integer
+   /// with 128 bits.
+   amount: string,
+   /// The valid NEAR account indicating the Fungible Token contract.
+   contract_id: string
+}
+
+
+
+
+/****************/
+/* VIEW METHODS */
+/****************/
+
+/// Deprecated - use `get_key_information` instead. This method provides backwards compatibility with previous implementations.
+/// Returns the $yoctoNEAR amount associated with a given public key 
+/// Panics if the key does not exist. 
+/// The Public key must be in a binary format with base58 string serialization with human-readable curve. The key types currently supported are secp256k1 and ed25519. Ed25519 public keys accepted are 32 bytes and secp256k1 keys are the uncompressed 64 format.
+function get_key_balance(key: string) -> string
+
+/// Returns the KeyInfo associated with a given public key
+/// Panics if the key does not exist
+/// The Public key must be in a binary format with base58 string serialization with human-readable curve. The key types currently supported are secp256k1 and ed25519. Ed25519 public keys accepted are 32 bytes and secp256k1 keys are the uncompressed 64 format.
+function get_key_information(key: string) -> KeyInfo
+
+/******************/
+/* CHANGE METHODS */
+/******************/
+
+/// Transfer all assets linked to the signer’s public key to an *existing* NEAR account. 
+/// If the transfer fails for whatever reason, it is up to the smart contract developer to 
+/// choose what should happen. For example, the contract can choose to keep the assets 
+/// or send them back to the original linkdrop creator.
+/// 
+/// Requirements:
+/// * The predecessor account *MUST* be the current contract ID.
+/// * The assets being sent *MUST* be associated with the signer’s public key.
+/// * The assets *MUST* be sent to the `account_id` passed in.
+///
+/// Arguments:
+/// * `account_id` the account that should receive the linkdrop assets.
+///
+/// Returns `true` if the claim was successful meaing all assets were sent to the `account_id`.
+function claim(account_id: string) -> Promise<boolean>
```

Comment:
This does not match the [near-linkdrop contract implementation](https://github.com/near/near-linkdrop/blob/4d2031b8fbb64973f0d0fdb64de75d635acb673d/src/lib.rs#L79-L96), which [returns](https://explorer.near.org/transactions/HeRgemAugcjXombjvuN8hvHnKyvubAfKtgL5bqQgeLw5) an empty string result on `claim` since it defers the transaction result to the instantiated TRANSFER receipt. Given that there is not so much value in the boolean (why would you return `false` instead of failing the function call?), I think there is no reason to break the compatibility.

P.S. On the other hand, `create_account_and_claim` [returns](https://explorer.near.org/transactions/AabcikcF9sGrKUtWskHWHUfgRRUkaRqzSsaias2KviEQ) a boolean.

Reply:
This is a tricky one... On the Keypom protocol, we do manual bookkeeping of an access key's allowance since this information isn't exposed in the trie. This is so that we can refund the funder for any unspent allowance. This bookkeeping won't work if the transaction fails since the state changes will be reverted.

I'm unsure as to how to proceed..

Reply:
Well, indeed, it is a tough compatibility issue now. I have a solution, but I cannot say I love it.

You can still return an error without crashing the current receipt - you crash a cross-contract receipt. I.e. `claim` function would update the storage, and initiate a cross-contract call that will crash. It is ugly and requires more gas for the cross-contract call in case of an error, but you should have enough gas to do that as you don't call the transfers and don't call the callback handler.

Given there is a misalignment on the near-linkdrop contract between the `claim` and `create_account_and_claim` calls, I would consider updating the `claim` implementation by adding the callback that will return `boolean` (you will want to reach out to wallets to check if this change affects them)

Reply:
@mattlockyer It seems that you [have a strong opinion on maintaining 100% compatibility with the near-linkdrop contract](https://github.com/near/NEPs/pull/452/files#r1171519981). What do you think about this breaking change?

@BenKurrek Given that panic is always possible (e.g. not enough gas), it is fair to expect that clients must also implement handlers for those (in addition to the boolean value). The problem is that the current clients only implemented the handlers for panics, and not for booleans, but once the standard is published, we can share it with the implementors to update their code.

Reply:
Isn't the existing contract upgradeable? Do we have a list of products which adopted it? What is the risk of breaking the compatibility now? If it's only a handful and we can easily update them, then breaking the compatibility for having a long term benefits is OK.

Reply:
> Given that panic is always possible (e.g. not enough gas), it is fair to expect that clients must also implement handlers for those (in addition to the boolean value). 

Well, if a client wants to handle the panic, then she/he will need to implement the handler anyway.

Reply:
The contract is upgradable, so we can change the behavior, but it will require a callback handler after the transfer to just return `true` or `false`.
The risk is that UI will break when claim failed without panic and reported `false`, but I think it is minor, and can be easily fixed.

Reply:
Actually, thinking about it from the client perspective, I think it is more beneficial to receive an object with the details about the claimed assets instead of just `true`. This would make things much more ergonomic in my opinion. Here is an example return value:

```js
{
  "status": "success", // or "error"
  // Even with "error" it might be worth providing "transferred_assets" as there might be partial transfers, since transfers are not atomic
  "transferred_assets": { // I would consider renaming `KeyInfo` struct into `LinkdropAssets` struct, and re-use it here
    "yoctoNEAR": "1",
    "nft_list": [],
    "ft_list": [{ /* ... */}]
  }
}
```

Reply:
I like the idea of returning a more complex type. It would also allow a use case for partial claims.
BTW, instead of `{status, transferred_assets}`, how about returning `Option<TransferredAssets>`?

Reply:
@frol @robert-zaremba Partial transfers are tricky, it's gonna be hard to manage on the contract side: what is the state of `KeyInfo` is a case of partial transfers, and what the `"status": "success"` would mean in that case, and what is the plan B, should the user claim the same key again?

I propose explicitly stating (in the NEP) that either all assets are transferred, or the claim failed. What do you think?

Reply:
I think it depends on the implementation, and the standard should allow for both cases. It's easy to handle when we allow return type. There could be 2 reasons for partial claim:
1. it's too expensive to do in a single transaction (too much gas to transfer too many tokens)
2. it requires 2 steps for whatever reason
3. not enough funds.

Implementation can adopt to each of the cases. Eg, in 3. the implementation can do a partial claim and remove the record all together. If we adopt `status`, then it could be more explicit by adding an enum `done | contine | ...`.

Reply:
@robert-zaremba extending the status enumeration looks like a good compromise. What if to return this new status *only*? User can always call a view method to see more details about Linkdrop assets? Of course, you can argue about saving a 1 view method call but it simplifies the NEP on the other hand.

Reply:
By the way, `String` would be okay-ish type (a bit ugly though) to make the NEP standard more agile to different ways of implementation. In this case, I suppose it's gonna be compatible with [near-linkdrop](https://github.com/near/near-linkdrop/blob/4d2031b8fbb64973f0d0fdb64de75d635acb673d/src/lib.rs#L79-L96)

Reply:
I would *prefer* to leave the success checks to the client as it's much simpler. In the case of FTs & NFTs, they will know what assets should be transferred and can just query the claimed account's tokens to ensure that things were successful. The only thing that may be a bit tricky is checking for the success of simple $NEAR drops (when there's no account creation involved).

For this reason, I propose 3 options:
- return a simple promise with no specific type as @fadeevab mentioned above (my favourite option)
- return a boolean indicating whether *all* transfers were successful. If anything was not successful, the client can do some checks to see what went wrong.
- return a boolean indicating whether or not the $NEAR was successfully transferred and leave the other checks up to the client.

Reply:
@fadeevab The cons of the current near-linkdrop approach (`void` on ok & `panic` on error) is that contract cannot save anything in its state when it panics (as the receipt execution is reverted), and while near-linkdrop does not need it, keypom wants to save something in its state while still signaling that the claim was not successful.

There is a workaround: make the necessary changes and submit a cross-contract call to a method (on itself) that will panic just to signal the failure (in NEAR, only receipt execution is atomic, so the changes committed in the `claim` function won't be reverted), but it will require more gas only to signal the failure, and will also look a bit cumbersome.

Another con is that `claim` and `create_account_and_claim` will have different ways to signal errors (near-linkdrop uses panic on `claim` and boolean on `create_account_and_claim`)

Reply:
@frol 
_The question 1: does this NEP suppose to consider/solve this particular problem (signaling a failure with boolean avoiding panic)?_ 
_The question 2: do we want to standardize different return types `claim()->void` and `create_account_and_claim()->bool` achieving backward compatibility with `near-linkdrop`?_ 

Around the first point: signaling a failure with boolean without panicking looks like a gross edge case for the NEAR contracts overall. If the developer needs to implement the mentioned "edge case" then the mentioned workaround is the way to go.

Anyways, depending on answers, we have 4 binary options for return types:

1. `void/void`: If both answers are "no", then `claim()->void`, `create_account_and_claim()->void` would be the way to go.
2. `bool/bool`: If solving the first problem is crucial ("yes"), but backward compatibility is not ("no"), then  `claim()->bool`, `create_account_and_claim()->bool`.

(and there are 2 more binary options of course :) )

Reply:
I am in favor for returning the `KeyInfo`. Can be in a `Promise` or in a `Result`.

Reply:
> keypom wants to save something in its state while still signaling that the claim was not successful

What they want to save? Would help to understand the issue

Reply:
@robert-zaremba There was a message from Ben above: https://github.com/near/NEPs/pull/452#discussion_r1134621955

Reply:
BTW, I still think we should consider advocating a `Result<_,_>` type (of the last receipt) which _will not_ trigger `panic`, but may signal transaction failure.
Panics / reverts are kind of a standard in the Ethereum synchronous execution model, but I think we should do more research about pros/cons for such approach.

Reply:
After further discussion, I propose to keep things simple and keep `boolean` as the return value of `claim` and `create_account_and_claim` functions. Here is my reasoning:

* This NEP is a reflection of the existing ecosystem and breaking it is not in our interest
* Errors in `claim` method are rare as usually dApps check if the claim is still available before submitting the transaction
* Having `boolean` return value for `claim` aligns it with the `create_account_and_claim`, which uses `boolean` as its return value since day one

This was the only remaining concern blocking this NEP, and @fadeevab @abacabadabacaba agreed that we don't need to block it any longer.